Black Formatting, Type Hints, Works on Windows

wallstop opened the pull request:
# Overview
This PR does a few things, I can split them into separate PRs if you're interested in taking any of them.

&check; Update gitignore to take into account Jetbrains IDEs (I used pycharm to edit this)
&check; Apply [Black](https://github.com/psf/black) for deterministic formatting
&check; Apply type hints to all types that had ambiguities, now you know for sure what everything is and your IDE will yell at you (if you have it configured) when it detects a violation
&check; Added safter directory deletion. This could trigger [symlink attacks](https://docs.python.org/3/library/shutil.html#shutil.rmtree), since you're both downloading and then executing rmtree on arbitrary files, which could be symlinks.
&check; Added more robust error handling around deleting files, specifically around the deletion of .git files on windows

cgoldberg commented:
Thanks! I just pushed some new updates (see #5) that cause conflicts with your branch. Can you merge my latest version into yours and fix the conflicts? 

After that, I'll review your changes and probably accept all of them. 

I actually ran black already, but I had it ignore single quotes and use a longer line length... but I don't really care and am fine with the defaults. 

I'll probably setup tox and black for easier linting and add some unit tests soon.

I also published another release to PyPI... I'll do another once I get your changes in.

wallstop commented:
> Thanks! I just pushed some new updates (see #5) that cause conflicts with your branch. Can you merge my latest version into yours and fix the conflicts?
> 
> After that, I'll review your changes and probably accept all of them.
> 
> I actually ran black already, but I had it ignore single quotes and use a longer line length... but I don't really care and am fine with the defaults.
> 
> I'll probably setup tox and black for easier linting and add some unit tests soon.
> 
> I also published another release to PyPI... I'll do another once I get your changes in.

Nice, I rebased latest master and changed all typing to optional (for maximum portability). Tested with and without token, things appear to be working.

wallstop commented on githubtakeout.py:
```diff
@@ -4,100 +4,158 @@
-    return new_url
+        netloc = f"{username}:{token}@{url_parts.netloc}"
+        url_parts = url_parts._replace(netloc=netloc)
+        return cast(str, urllib.parse.urlunparse(url_parts))
```
For some reason my type hinter was telling me this was returning a byte string literal, which isn't true.

cgoldberg commented:
Thanks again for the contribution, but I am not going to merge the PR as-is.

I made a few changes to the master branch that address some of these issues. I renamed `zip` and `tar` so they don't shadow builtins, and a few other minor things. (see #7)

I also took a shot at fixing the issue you mentioned getting permission errors using `rmtree()`.  The docs have a nice way of handling that, so I added something similar:

https://docs.python.org/3/library/shutil.html#rmtree-example

Unfortunately, this requires Python 3.12+. Rather than making it backwards compatible (which isn't that hard), I just dropped support for Python < 3.12. (It will raise an error if you run it on older versions).

Can you try it on Windows and see if it works, or if it still causes errors?

I didn't remove `rmtree()` because I don't quite understand the concern about following symlinks. Is it that there might be symlinks in a repo pointing elsewhere in the filesystem that would get wiped out?

Most of the other changes in the PR were the formatting and import changes, which I didn't care for.

The only outstanding thing I didn't address is the type hints. If you want to create a PR off my master branch with just the typng changes, I will merge that.


Again, sorry for not accepting the PR... but I really do appreciate the effort.

wallstop commented:
Oh hey, no problem, if this PR was able to improve the code in any way, that's all that matters.

RE: shutil.rmtree, I found this old bug: https://github.com/python/cpython/issues/48739. From [the docs](https://docs.python.org/3/library/shutil.html):

---
Note

On platforms that support the necessary fd-based functions a symlink attack resistant version of [rmtree()](https://docs.python.org/3/library/shutil.html#shutil.rmtree) is used by default. On other platforms, the [rmtree()](https://docs.python.org/3/library/shutil.html#shutil.rmtree) implementation is susceptible to a symlink attack: given proper timing and circumstances, attackers can manipulate symlinks on the filesystem to delete files they wouldn’t be able to access otherwise. Applications can use the [rmtree.avoids_symlink_attacks](https://docs.python.org/3/library/shutil.html#shutil.rmtree.avoids_symlink_attacks) function attribute to determine which case applies.
---

So the code that I added to replace the use of rmtree was meant to both address that issue and also gracefully handle windows-related errors. If you want this to be "production-ready" and  you really want to use rmtree, recommend maybe checking the `avoids_symlink_attacks` property and falling back to a safe impl if it is false. Or just use a safe version always. Otherwise a well-formed payload could cause data loss using this tool. Besides that, the main reason I changed rmtree was to get around it failing on windows to delete files in the .git directory, which causes the program to blow up on the first repo deletion.

Anyways, since you now require python 3.12, I can no longer test the code, as I'm locked in to a version of python 3.10 for unrelated projects and I really don't want to deal with managing multiple versions of python on my path.

Cheers!

cgoldberg commented:
Have you tried pyenv for managing Python versions? It works really well. I juggle between Python versions all day. There is Windows version: https://github.com/pyenv-win/pyenv-win

Anyway, thanks again .. your code and ideas helped me improve it.

wallstop commented:
Thanks for the pyenv link. Latest version appears to work on Windows, nice, thanks.

I was considering created a PR with your Black preferences in the pyproject.toml so it could be reproducible, but it appears that Black can only be configured to either *not* apply quote formatting (keep existing quotes, regardless of format), or turn everything into their style, which is double quotes, [by design](https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#labels-strings). History [1](https://github.com/psf/black/issues/118), [2](https://github.com/psf/black/issues/594). Since your preference is single quotes, that idea's a no-go. If you're invested in Black as a tool, might be worth considering switching string styles. But if you're not, or if you prefer tools like ruff that give more control to you, then keep on doing you. Thanks for the tool.
